lib/web_ui/lib/src/engine/skwasm/skwasm_impl/renderer.dart

```diff
@@ -330,19 +330,41 @@ class SkwasmRenderer implements Renderer {
     int? targetHeight,
     bool allowUpscaling = true
   }) {
+
+    if (targetWidth != null) {
+      assert(allowUpscaling || targetWidth <= width);
+    }
+    if (targetHeight != null) {
+      assert(allowUpscaling || targetHeight <= height);
+    }
+    // Skia can copy these pixels into a JS clamped array a lot faster with
+    // ffi than js_interop ideally we could skip a temporary heap allocation,
+    // but until there is a fast Uint8Array copy path, this is much faster.
     final SkwasmImage pixelImage = SkwasmImage.fromPixels(
       pixels,
       width,
       height,
       format
     );
-    final ui.Image scaledImage = scaleImageIfNeeded(
-      pixelImage,
-      targetWidth: targetWidth,
-      targetHeight: targetHeight,
-      allowUpscaling: allowUpscaling,
-    );
-    callback(scaledImage);
+
```

Comment:
I'm having trouble understanding the purpose of this change here. We already have an image at this point, and it appears the new code change explicitly is pulling the pixels back out of that image and then creating an image bitmap from those pixels, which basically incurs another back and forth between the CPU and GPU. I can't imagine this would actually be faster than the code before this change. Can you elaborate on why we're doing it this way?

Reply:
The goal of the change was the move the SkImage out of the CanvasKit heap, so that it lives in the normal browser GC space instead of inside CanvasKit. What we are seeing is that it's very easy for libraries to run the canvaskit heap out of memory even though the browser has plenty of memory.  The problem is that once the CanvasKit heap runs out of memory, flutter becomes very unstable, hangs, or crashes, and it is not recoverable. The best way to prevent this instability is to keep data out of the CanvasKit heap when possible. Since image data can very quickly exhaust the entire available heap space if it is copied into CanvasKit, this change prevents that from happening with this API.

Here's an example project that illustrates the problem by loading some 4k data:

https://github.com/jezell/mem_test

![image](https://github.com/user-attachments/assets/192f3ffe-9711-4e69-93c6-fcd6f6b1a29e)
![image](https://github.com/user-attachments/assets/5b6a4606-0e01-4b9b-a0cf-6234b6bac609)

If instead of moving the buffer into the Skia heap, we bind the texture to something outside the heap, instead of using 25% of the available heap, we use 0% of the heap. I believe this is a big win for stability with packages like pdfrx that to load a lot of image data asynchronously. At the moment, libraries like pdfrx are resorting to things like this to try to prevent the process from blowing up:

https://github.com/espresso3389/pdfrx/blob/27228e8b64dd6444c5c47b47f5c1b0f2e05ae3b6/lib/src/widgets/pdf_viewer.dart#L1106

But that doesn't actually work that well in practice because async processes that create the images can easily step on each other and run canvaskit out of memory temporarily, at which point the renderer gets into a bad an unrecoverable state.

Reply:
The new code still allocates all the pixels on the skia heap in `SkwasmImage.fromPixels`. So it seems like maybe the reason the old code is bloating memory is because the scaled output image is internally holding on to the unscaled raw pixel data. If I'm understanding correctly, this means that the benefit to the new code is that the allocation of the raw pixel bytes on the skia heap is temporary rather than tied to the lifetime of the returned image.

If that's the case, I think we can probably have our cake and eat it too. That is, avoid the long-term allocation of those bytes and also avoid more CPU-to-GPU-to-CPU copies. I think this could work as follows:

* Draw the `pixelImage` to a `Picture` object (that's the correct scaled size)
* Use the `renderer.surface.renderPictures` API to render that picture and get an image bitmap out of it (see https://github.com/flutter/engine/blob/main/lib/web_ui/lib/src/engine/skwasm/skwasm_impl/surface.dart#L92C26-L92C26)
* Use `createImageFromTextureSource` on the image bitmap that is returned from that (like you're already doing below).

That way, I think we will achieve the same thing this does, but a lot more efficiently. Once we've uploaded the image data to the GPU it can essentially stay there (if the browser does the right thing).

Let me know if that makes sense and if that works for the thing we're trying to achieve here.

Reply:
Yeah I didn't like the temporary use of the skwasm heap here either. Originally I had it just creating the ImageData directly, but was running into some performance issues from the js_interop checks on copying all the data over.  Might be able to avoid that with the suggestion from @mkustermann to use ```jsArray.toDart[index]``` for the fast path.